### ISSUE TITLE
About Pretraining EEG Dataset

### ISSUE THREAD

**Issue by leofan90** [bbaaii/DreamDiffusion#9]
Hello! I'm running EEG pretraining with code stageA1_eeg_pretrain.py. I'm wondering do you provide the Pretraining EEG Dataset? I saw the path in this code (../dreamdiffusion/datasets/mne_data/) doesn't  include in the README.md File path Description. I'm looking forward to your reply. Thank you!

**Comment by laddg95** [comment]
Had you resolve it?

**Comment by Zoe-Wan** [comment]
I have the same problem.

**Comment by laddg95** [comment]
I know what's going on. The `mne_data` here actually comes from the `\datasets\eeg_5_95_std.pth` file, but you need to extract the EEG data in the file, convert each tensor to ndarray and save it as one document. In my case, I named the file `i.npy`, where i is the index of each tensor, and now the program runs successfully. In addition, the shape of each tensor is (128, L), and L is not equal. The `__getitem__` method in the `eeg_pretrain_dataset` class will unify L.

**Comment by Zoe-Wan** [comment]
I extract the numpy arrays from the pth file successfully! Thanks to @laddg95. 
But I still think this amount of data ( 11952 EEG records, for each has size 128*500) is not enough for pretraining task ( the author said it should be around "120,000 EEG data samples from over 400 subjects with channel ranges from 30 to 128" )

**Comment by laddg95** [comment]
Your suspicions should be correct, after all the pre-training data and test data should not be the same. Maybe the author wants us to use our own pre-training data. I noticed that in other issues, others provided links to other data sources, which may be used as pre-training data. I don't have time to verify it yet. If your experiment is successful, you can reply to me here, I will be very grateful! Thanks!

**Comment by moyu2023101** [comment]
@laddg95 You can still download it now \Datasets  eeg_ 5_ 95_ Is the std.pth file or other similar pre trained EEG checkpoint? Or in which directory it is located, I cannot find it. If you could provide me with a link, I would greatly appreciate it.

**Comment by laddg95** [comment]
Please see the `README` section
[https://studentiunict-my.sharepoint.com/personal/concetto_spampinato_unict_it/_layouts/15/onedrive.aspx?ga=1&id=%2Fpersonal%2Fconcetto%5Fspampinato%5Funict%5Fit%2FDocuments%2Fsito%5FPeRCeiVe%2Fdatasets%2Feeg%5Fcvpr%5F2017%2Fdata](url)

**Comment by moyu2023101** [comment]
The problem has been resolved!!! Thanks to @laddg95 !

**Comment by miao-zhuyue** [comment]
> I know what's going on. The `mne_data` here actually comes from the `..\datasets\eeg_5_95_std.pth` file, but you need to extract the EEG data in the file, convert each tensor to ndarray and save it as one document. In my case, I named the file `i.npy`, where i is the index of each tensor, and now the program runs successfully. In addition, the shape of each tensor is (128, L), and L is not equal. The `__getitem__` method in the `eeg_pretrain_dataset` class will unify L.

Thanks!!!!

**Comment by SunshineXiang** [comment]
> I know what's going on. The `mne_data` here actually comes from the `..\datasets\eeg_5_95_std.pth` file, but you need to extract the EEG data in the file, convert each tensor to ndarray and save it as one document. In my case, I named the file `i.npy`, where i is the index of each tensor, and now the program runs successfully. In addition, the shape of each tensor is (128, L), and L is not equal. The `__getitem__` method in the `eeg_pretrain_dataset` class will unify L.

I have downloaded this file, but I do not know how to make npy file

**Comment by SunshineXiang** [comment]
> I know what's going on. The `mne_data` here actually comes from the `..\datasets\eeg_5_95_std.pth` file, but you need to extract the EEG data in the file, convert each tensor to ndarray and save it as one document. In my case, I named the file `i.npy`, where i is the index of each tensor, and now the program runs successfully. In addition, the shape of each tensor is (128, L), and L is not equal. The `__getitem__` method in the `eeg_pretrain_dataset` class will unify L.

I have downloaded this file, but I do not know how to make npy file.

**Comment by SunshineXiang** [comment]
RuntimeError: CUDA out of memory. Tried to allocate 2.00 GiB (GPU 0; 23.65 GiB total capacity; 19.50 GiB already allocated; 410.06 MiB free; 21.39 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

**Comment by Venchy-he** [comment]
@Zoe-Wan Have you finished this experiment?

**Comment by Tom345345** [comment]
@laddg95 .Hello, did you run successfully?  I can not find  these documents——" generation/checkpoint_best.ptn  and  eeg_pretain/checkpoint.pth ".  Can you share them? Thank you very much。

**Comment by Zoe-Wan** [comment]
> @Zoe-Wan Have you finished this experiment?

Yes, I have used the extracted data to pretrain, and I have tried some params which is mentioned in the paper ( such as mask ratio, patch size etc. ). I find as the mask ratio is higher, the reconstruction performance is better . I got 0.55 average correlation (using the eval method in the code) when use mask ratio 0.75 and patch size 4 . And the reconstruction result is like the pic below.
![reconst-08-10-2023-15-59-56](https://github.com/bbaaii/DreamDiffusion/assets/36704065/aa1f3129-b24f-43bb-8fdb-2e38faf92747)

But unfortunately I do not have enough GPU memory ( just like @SunshineXiang  ) to continue this work. And even if I have enough GPU memory, the pth datasat it gives is for finetune, I still can not continue. And apparently,  11952 EEG records is not enough for pretrain, because it does not split the train dataset and the test dataset. Overfitting a little dataset does not make sense, so I still suggest that you find another dataset to do this pretraining task (do not forget preprocessing).

**Comment by Zoe-Wan** [comment]
What's more, I suppose that  segmenting EEG data into alpha waves, beta waves, theta waves, etc., based on psd should be a good idea to preprocess.

**Comment by miao-zhuyue** [comment]
> > 你完成这个实验了吗？
> 
> 是的，我已经使用提取的数据进行预训练，并且我尝试了一些论文中提到的参数（例如掩模比例，补丁大小等）。我发现随着掩模率的提高，重建性能越好。我得到了 0.55 的平均相关性（在代码中使用 eval 方法），同时使用掩码比 0.75 和补丁大小 4 .重建结果如下图所示。 ![Reconst-08-10-2023-15-59-56](https://user-images.githubusercontent.com/36704065/275707056-aa1f3129-b24f-43bb-8fdb-2e38faf92747.png)
> 
> 但不幸的是，我没有足够的 GPU 内存（就像 ）来继续这项工作。即使我有足够的 GPU 内存，它给出的 pth 数据集也是用于微调的，所以我仍然无法继续。显然，11952 个脑电图记录不足以进行预训练。因为它没有拆分训练数据集和测试数据集，所以在一个小数据集中过度拟合是没有意义的，所以我还是建议你找另一个数据集来做这个预训练任务（不要忘记预处理）。

Last month, I found that you have asked about pretrains/generation/checkpoint_bset.pth, this file, had you solved this problem?Does this file need to be trained by yourself? If so, may I ask in which file?

**Comment by Zoe-Wan** [comment]
> > > 你完成这个实验了吗？
> > 
> > 
> > 是的，我已经使用提取的数据进行预训练，并且我尝试了一些论文中提到的参数（例如掩模比例，补丁大小等）。我发现随着掩模率的提高，重建性能越好。我得到了 0.55 的平均相关性（在代码中使用 eval 方法），同时使用掩码比 0.75 和补丁大小 4 .重建结果如下图所示。 ![Reconst-08-10-2023-15-59-56](https://user-images.githubusercontent.com/36704065/275707056-aa1f3129-b24f-43bb-8fdb-2e38faf92747.png)
> > 但不幸的是，我没有足够的 GPU 内存（就像 ）来继续这项工作。即使我有足够的 GPU 内存，它给出的 pth 数据集也是用于微调的，所以我仍然无法继续。显然，11952 个脑电图记录不足以进行预训练。因为它没有拆分训练数据集和测试数据集，所以在一个小数据集中过度拟合是没有意义的，所以我还是建议你找另一个数据集来做这个预训练任务（不要忘记预处理）。
> 
> Last month, I found that you have asked about pretrains/generation/checkpoint_bset.pth, this file, had you solved this problem?Does this file need to be trained by yourself? If so, may I ask in which file?

Yeah, you should train it by yourself, the author has not shared the file yet. But just like I said, I can not finetune the model, so I do not have the ckpt file neither.  If you want to try it, you can organize the file just like what readme file says.

**Comment by Tom345345** [comment]
@Zoe-Wan
Hello, I see in the above information that you have successfully run the code. I am a novice. Do these two files named (pretrains/generation/checkpoint_bset.pth 
and pretrains/eeg_pretain/checkpoint_bset.pth) really affect the code operation? Could you share these two files? I would appreciate it. What I said is not correct, please give me your advice.

**Comment by Tom345345** [comment]
@Zoe-Wan
Hello, I see in the above information that you have successfully run the code. I am a novice. Do these two files named (pretrains/generation/checkpoint_bset.pth
and pretrains/eeg_pretain/checkpoint_bset.pth) really affect the code operation? Could you share these two files? I would appreciate it. What I said is not correct, please give me your advice.

**Comment by bbaaii** [comment]
You can directly download this data from the MOABB project   https://github.com/NeuroTechX/moabb   and use their respective processing code for operations like filtering. Finally, save them as npy files.

**Comment by caltexs** [comment]
> > > > 你完成这个实验了吗？
> > > 
> > > 
> > > 是的，我已经使用提取的数据进行预训练，并且我尝试了一些论文中提到的参数（例如掩模比例，补丁大小等）。我发现随着掩模率的提高，重建性能越好。我得到了 0.55 的平均相关性（在代码中使用 eval 方法），同时使用掩码比 0.75 和补丁大小 4 .重建结果如下图所示。 ![Reconst-08-10-2023-15-59-56](https://user-images.githubusercontent.com/36704065/275707056-aa1f3129-b24f-43bb-8fdb-2e38faf92747.png)
> > > 但不幸的是，我没有足够的 GPU 内存（就像 ）来继续这项工作。即使我有足够的 GPU 内存，它给出的 pth 数据集也是用于微调的，所以我仍然无法继续。显然，11952 个脑电图记录不足以进行预训练。因为它没有拆分训练数据集和测试数据集，所以在一个小数据集中过度拟合是没有意义的，所以我还是建议你找另一个数据集来做这个预训练任务（不要忘记预处理）。
> > 
> > 
> > Last month, I found that you have asked about pretrains/generation/checkpoint_bset.pth, this file, had you solved this problem?Does this file need to be trained by yourself? If so, may I ask in which file?
> 
> Yeah, you should train it by yourself, the author has not shared the file yet. But just like I said, I can not finetune the model, so I do not have the ckpt file neither. If you want to try it, you can organize the file just like what readme file says. ( This comment may be replied to #5 )

excuse my stupidness, but couldn't someone do all the training and then upload the ckpt file for us to use? Or why doesnt the author do that? Not a programmer, so sorry if this is a stupid question.

**Comment by NZW666666** [comment]
> e

这啥意思啊哥

**Comment by NZW666666** [comment]
这段代码中的路径（../dreamdiffusion/datasets/mne_data/） 不包含在 README.md 文件路径说明中，有没有大佬帮帮忙，上面大佬说的看不太懂，谢谢

**Comment by KrishKrosh** [comment]
> You can directly download this data from the MOABB project https://github.com/NeuroTechX/moabb and use their respective processing code for operations like filtering. Finally, save them as npy files.

From my understanding, the MOABB dataset has multiple different datasets and processing steps for different paradigms. Can you let us know what specific datasets, processing steps, and params you used?

**Comment by chenjunxia06** [comment]
@Zoe-Wan  Hello, how can I get from ".. datasetseeg_5_95_std.pth" file? I am a novice.  Please give me your advice.I would appreciate it.

**Comment by chenjunxia06** [comment]
@Zoe-Wan Hello, how can I get the numpy from ".. datasetseeg_5_95_std.pth" file? I'm a newbie. Please give me your advice. I would be grateful.

**Comment by qiaoyub** [comment]
> 我成功地从 pth 文件中提取了 numpy 数组！感谢 .但我仍然认为这个数据量（11952 个脑电图记录，每个都有 128*500 的大小）不足以完成预训练任务（作者说它应该在“来自 400 多个受试者的 120,000 个脑电图数据样本，通道范围从 30 到 128 不等”）

Hello，about the pth extraction data, and then EEG pre-training operation of the specific process of how to work, I need your help very much, thank you very much！！

**Comment by qiaoyub** [comment]
> 这段代码中的路径（../dreamdiffusion/datasets/mne_data/） 不包含在 README.md 文件路径说明中，有没有大佬帮帮忙，上面大佬说的看不太懂，谢谢

请问你解决这个问题了吗

**Comment by usernamenotavaila** [comment]
> le `i.npy`

It works.Thank you very much for solving the problem.

**Comment by usernamenotavaila** [comment]
> I know what's going on. The `mne_data` here actually comes from the `..\datasets\eeg_5_95_std.pth` file, but you need to extract the EEG data in the file, convert each tensor to ndarray and save it as one document. In my case, I named the file `i.npy`, where i is the index of each tensor, and now the program runs successfully. In addition, the shape of each tensor is (128, L), and L is not equal. The `__getitem__` method in the `eeg_pretrain_dataset` class will unify L.

It works.Thank you very much for solving the problem

**Comment by mahui-HUE** [comment]
> I know what's going on. The `mne_data` here actually comes from the `..\datasets\eeg_5_95_std.pth` file, but you need to extract the EEG data in the file, convert each tensor to ndarray and save it as one document. In my case, I named the file `i.npy`, where i is the index of each tensor, and now the program runs successfully. In addition, the shape of each tensor is (128, L), and L is not equal. The `__getitem__` method in the `eeg_pretrain_dataset` class will unify L.

What EEG datasets have you downloaded?thank you!

**Comment by mahui-HUE** [comment]
> > You can directly download this data from the MOABB project https://github.com/NeuroTechX/moabb and use their respective processing code for operations like filtering. Finally, save them as npy files.
> 
> From my understanding, the MOABB dataset has multiple different datasets and processing steps for different paradigms. Can you let us know what specific datasets, sampling rate, processing steps, and params you used?

Pls have you solved it, which datasets were downloaded?

**Comment by mahui-HUE** [comment]
> 我知道是怎么回事。`mne_data`这里的 实际上是来自`..\datasets\eeg_5_95_std.pth`文件，但是你需要提取文件中的脑电数据，将每个张量转换为 ndarray 并将其保存为一个文档。在我的例子中，我将文件命名为`i.npy`，其中 i 是每个张量的索引，现在程序成功运行。另外，每个张量的形状都是 (128, L)，并且 L 不相等。类`__getitem__`中的方法`eeg_pretrain_dataset`将统一 L。

感谢你的提醒，我已经成功运行EEG预训练命令。
